### ISSUE TITLE
ConnectedDatagramPort uses deprecated `loseConnection` API when a connection fails

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3425]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3425|
|Type|defect|
|Created|2008-09-09 15:48:11Z|
|Branch|https://github.com/twisted/twisted/tree/use-stop-listening-3425|

This results in a deprecation warning about using `stopListening` instead of `loseConnection` any time a `connectUNIXDatagram` attempt fails.

<details><summary>Searchable metadata</summary>

```
trac-id__3425 3425
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_use_stop_listening_3425 branches/use-stop-listening-3425
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1220975291000000 1220975291000000
changetime__1267554859000000 1267554859000000
version__None None
owner__ 
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess commented|
|-|-|

(In [28074]) Branching to 'use-stop-listening-3425'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess commented|
|-|-|

(In [28093]) ConnectedDatagramPort now uses stopListening when a connection fails
instead of the deprecated loseConnection. Adjust docstring formatting
in surrounding functions and add a unit test for this change in
behavior.

refs [#3425](https://github.com/twisted/twisted/issues/3425)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @jesstess|
|-|-|

The implementation change looks good to me.  I'd like to see the test added to `twisted.internet.test.test_unix` instead, though, and an assertion added to it.  It'd be nice if it just exercised the real implementation, too - and as a side effect, the lack of a warning being emitted will tells us that it isn't using the deprecated method.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess commented|
|-|-|

(In [28390]) Move test to twisted/internet/test/test_unix.py and don't use a dummy
stopListening.

refs [#3425](https://github.com/twisted/twisted/issues/3425)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [28393]) Skip on platforms that don't support UNIX sockets.

refs [#3425](https://github.com/twisted/twisted/issues/3425)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @jesstess|
|-|-|

  1. In `twisted/internet/test/test_unix.py`, the line `test_usesStopListening = skip` looks like a mistake.
  1. It's not clear by looking at `test_unix.py`, I guess, but this module is intended for blackbox tests.  I suggest `test_posixbase.py` for these tests.
  1. I think you can skip the `flushWarnings`.  It's enough that the warning isn't emitted when the tests are run.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess commented|
|-|-|

(In [28503]) Move test to test_posixbase.py.

refs [#3425](https://github.com/twisted/twisted/issues/3425)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess removed owner|
|-|-|

Addressed review comments 1-3 by exarkun above.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_12"></a>|@dreid set owner to @jesstess|
|-|-|

Hi jess!  This is great work.  Just a few comments.
1. The test-case-name should include twisted.internet.test.test_posixbase
2. And several lines in ConnectedDatagramPort are longer than 80 characters and should be reformatted.
3. The test method in test_posixbase is called usesStopListening but it only tests that we're not calling loseConnection.
 * It should be renamed to indicate that it is testing that we no longer call loseConnection
 * A new test method should be added that ensures that we are using stopListening

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_13"></a>|@dreid commented|
|-|-|

Same review, proper formatting!

Hi jess! This is great work. Just a few comments.

 1. The test-case-name should include twisted.internet.test.test_posixbase
 2. And several lines in ConnectedDatagramPort? are longer than 80 characters and should be reformatted.
 3. The test method in test_posixbase is called usesStopListening but it only tests that we're not calling loseConnection.
  A. It should be renamed to indicate that it is testing that we no longer call loseConnection
  B. A new test method should be added that ensures that we are using stopListening

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess commented|
|-|-|

(In [28519]) Split use of stopListening and disuse of loseConnection into separate tests.

refs [#3425](https://github.com/twisted/twisted/issues/3425)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess removed owner|
|-|-|

Thanks for the review, dreid!

Addressed review comments 1-3 by dreid above.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @jesstess|
|-|-|


```
+        @type reason: C{str}
+        @param reason: Why the connection failed.
```
 * reason is actually a `Failure` instance. Maybe the tests should be fixed to use a `Failure` instead of a string.
 * `test_connectionFailedDoesntCallLoseConnection` and `test_connectionFailedCallsStopListening` should probably have different docstrings.
 * `test_connectionFailedCallsStopListening` should use `assertEquals` instead of `failUnlessEquals`

This is fairly minor, so please merge once fixed. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_17"></a>|@jesstess commented|
|-|-|

(In [28618]) Address review points by therve.

refs [#3425](https://github.com/twisted/twisted/issues/3425)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_18"></a>|@jesstess set status to `closed`|
|-|-|

(In [28620]) Merge use-stop-listening-3425

Author: jesstess
Reviewers: exarkun, dreid, therve
Fixes: [#3425](https://github.com/twisted/twisted/issues/3425)

ConnectedDatagramPort now uses stopListening when a connection fails
instead of the deprecated loseConnection.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
